keps/sig-api-machinery/2161-apiserver-default-labels/README.md

```diff
@@ -0,0 +1,452 @@
+
+Matching a namespace by its name, on the other hand, is a more reliable way to whitelist namespaces as it's much easier to specifically allow only namespaces user control since user knows their own namespace names.
+
+#### Lack of permissions
+
+Users want to allow DNS traffic to coreDNS pods which are created in system generated namespaces like kube-system.
+Since these system-generated namespaces do not come up with labels, users may not be authorized to add labels to
+such namespaces and thereby accessing some essential services is not straight forward.
+
+#### Simplicity
+
+It is being reported that users find it difficult to work with mutation and validation webhooks as they are not
+able to select a single namespace to enforce those controllers. Similarly, it is quite common and intuitive for
+users to remember the names of the namespaces rather than labels associated with those namespaces while defining
+NetworkPolicies or writing admission controller configurations.
+
+Thus, regardless of wether such a user story is valid, it is clearly 'simpler' to implement (as are other similar stories) when a default mechanism for namespace selection is available.
+
+### Notes/Constraints/Caveats (Optional)
+
+* The name of the namespace must satisfy label value [requirements](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set).
+Currently, the name of the namespace must be a valid [DNS label](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names), and this requirement will not change going forward.
+
+### Risks and Mitigations
+
+* Every namespace gets bigger in size. We think this is inconsequential
+* Some user may already be using that label already. This is mitigated by the 
+fact that the whole "kubernetes.io" label prefix is [reserved](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set). However, since there
+is no validation of reserved label prefixes, an upgrade would mutate such a label (this is not a major concern).
+* any clients/applications expecting this label (i.e. for a NetworkPolicy), wouldnt see it, and thus they might not be able to find namespaces using this label technique.  In the case of network policies, this would result in traffic being blocked due to unavailable namespace labels.
+
+For this reason, we can do one alpha release where we log/event if we observe usage of this key that is not compatible.
+
+## Design Details
+
+This can be implemented by modifying the defaults.go and validation.go components for the namespace API, i.e., in the defaults.go file for api/core:
+
+<<[UNRESOLVED See https://github.com/kubernetes/kubernetes/pull/96968 where this evolves, but this example is functionally valid, modulo warnings ]>>
```

Comment:
My understanding of the discussion is that this is resolved.  Setting the value should be unconditional.  It's our label. No need to preserve the value.

The only open question is whether to do it silently in defaults or with a warning in strategy.  The warning seems appropriate to me, but we'll need to cover all create/update/get paths, which is slightly more complicated than defaults.go

If we don't do the warning, we can do it in defaults and just stomp the value.

I'm advocating for the warning, but only weakly.